### ISSUE TITLE
Make `<section/>` wrappers optional (for HTML)

### ISSUE THREAD

**Issue by Brixy** [jgm/djot#221]
Thank you so much for Djot. Djot will definitely end hopping between markdown and asciidoctor for me.

---

I, too, have been using Djot in production, and it’s a great experience.

From what I can tell the `<section/>` wrapper _can_ cause some troubles—in HTML context.

In [this post](https://github.com/jgm/djot/issues/213#issuecomment-1438780458) @jgm asks

> As for ugliness, I think it's no more ugly that {section=false} (though perhaps less explicit as to its purpose, at least for people who know English).
>
> What's the real-world case that is motivating this?

In all our webprojects we use these CSS rulesets:

```
* {
	margin: 0;
}

.flow > * + * {
	margin-top: 1.5em;
}
```

The `flow` class is added to `<main>` or some selected content wrappers.

The content within any automatically generated wrapper is not affected, of course.

Moreover `<section/>` is frequently used to group content ‘manually’ and visually, e.g. for zebra styles, alternating right-left/left-right aligned content etc.

All in all I think `<section/>` is rather a _layout_ than a _content_ tag (just like in [this HTML minifier](https://github.com/wilsonzlin/minify-html/blob/master/rust/common/spec/tag/whitespace.rs#L160)).

**Possible solutions:**

- Option to make `<section/>` wrappers optional
- Option to define a standard class name that is added to all Djot-generated `<section/>` tags (e.g. `flow`)

Thank you!

**Comment by jgm** [comment]
> All in all I think <section/> is rather a layout than a content tag (just like in [this HTML minifier](https://github.com/wilsonzlin/minify-html/blob/master/rust/common/spec/tag/whitespace.rs#L160)).

I think we're using `<section>` correctly, judging from the documentation [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section).

You can use a filter to remove the section entirely:

```js
return {
  section: (el) => {
    return el.children;
  }
}
```

Save as nosection.js and run with `djot --filter nosection.js`.

This will also remove the automatic identifiers, though.  Alternatively, you could add a special class:

```js
return {
  section: (el) => {
    el.attributes.class = "djot";
  }
}
```

If you're using djot as a library, you could also use the HTML renderer override documented on the readme, to render the section as something else...

**Comment by Brixy** [comment]
Thank you, John, for your answer and technical insights. Very interesting and helpful.

> I think we're using `<section>` correctly, judging from the documentation [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section).

Yes, of course: The generated HTML code is not only valid but also semantic.

Actually I forgot to mention one point: I was not talking about djot-to-HTML standalone files but about (rather complex) webprojects generated with an SSG using Djot as the markup language for the content. In these cases wrappers around the actual content _may_ cause problems and break the design, depending on the project.

Of course, it is always possible to adress these problems with CSS (but it gets trickier), and it’s good to know how to turn Djot’s wrapping functionalities of if needed.

Thanks again—and thank you so much for Djot!

**Comment by Brixy** [comment]
> One difference is that the example here doesn't have the topmost section wrapping h1 …

Good point. (It has always been a problem that HTML lacks some kind of title tag for the body. Using h1 was not the best idea and caused a lot confusion.)

**Comment by Gavin-Holt** [comment]
Hi, 

I am trying to "wrap" my markdown to make a tidy web page.

To keep it simple, I want my solution to work using both [djot.lua](https://github.com/jgm/djot.lua) and the windows executable from [jotdown](https://github.com/hellux/jotdown/releases).

This is my template `text` for *.djot files.

````html
```=html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
    <meta charset="utf-8"/>
    <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate"/>
    <meta http-equiv="Pragma" content="no-cache"/>
    <meta http-equiv="Expires" content="0"/>
    <link type="text/css" rel="stylesheet" href="normalize.css"/>
    <link type="text/css" rel="stylesheet" href="MutliMarkdown2.css"/>
    <script src="highlight.min.js"></script>
    <script src="sorttable.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
</head>
<body>
```

# This is a heading

some text

```=html
</body>
</html>
```
````

Running this through `jotdown.exe`  I get an unwanted `</section>` tag below `</html>` :

````html<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
    <meta charset="utf-8"/>
    <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate"/>
    <meta http-equiv="Pragma" content="no-cache"/>
    <meta http-equiv="Expires" content="0"/>
    <link type="text/css" rel="stylesheet" href="normalize.css"/>
    <link type="text/css" rel="stylesheet" href="MutliMarkdown2.css"/>
    <script src="highlight.min.js"></script>
    <script src="sorttable.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
</head>
<body>
<section id="This-is-a-heading">
<h1>This is a heading</h1>
<p>some text</p>
</body>
</html>
</section>
````

I get the same output using Lua `print(djot.render_html(djot.parse(text)))`.

Trying to add a blank line after "some text" does not work - it gets stripped out.

None of my browsers have complained - but it looks a bit wrong 🙁

Am I using djot incorrectly?

Kind Regards Gavin Holt

**Comment by jgm** [comment]
I would not recommend doing it this way.  Use djot to generate the main content and insert this into a template using a templating system of some kind.

**Comment by MichielioZ** [comment]
> I think we're using `<section>` correctly, judging from the documentation [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section).

If I'm writing `### Heading 3`, I'm not asking for a correct implementation of `<section>` though, I'm asking for a correct implementation of `<h3>`.  Sure, they are called [section heading elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/Heading_Elements) on MDN, but both MDN and the [official spec](https://html.spec.whatwg.org/multipage/sections.html#the-h1,-h2,-h3,-h4,-h5,-and-h6-elements) agree that headings have the same semantic meaning when using them direct or within `<section>`.

Wrapping a heading in `<section>` is opinionated, which is fine, but it would be nice to have the option to turn it off (or perhaps remove it and create syntax to create sections manually).
(Also, `<h1>` is wrapped in `<section>`, which isn't the case for the examples either on MDN or WWHATWG)

**Comment by dereuromark** [comment]
I added https://github.com/php-collective/djot-php/pull/266 for the PHP reference implementation for now.

One thing I noticed:
The section topic seems only relevant for root level, inside nested (e.g. blockquote, div, ..) there is no section wrapping anyway.
